emukit/bayesian_optimization/local_penalization_calculator.py

```diff
@@ -0,0 +1,93 @@
+                            output positive values only.
+        :param acquisition_optimizer: AcquisitionOptimizer object to optimize the penalized acquisition
+        :param model: Model object, used to compute the parameters of the local penalization
+        :param parameter_space: Parameter space describing input domain
+        :param batch_size: Number of points to collect in each batch
+        """
+        self.acquisition = acquisition
+        self.acquisition_optimizer = acquisition_optimizer
+        self.batch_size = batch_size
+        self.model = model
+        self.parameter_space = parameter_space
+
+    def compute_next_points(self, loop_state: LoopState) -> np.ndarray:
+        """
+        Computes a batch of points using local penalization.
+
+        :param loop_state: Object containing the current state of the loop
+        """
+
+        # Initialize local penalization acquisition
+        local_penalization_acquisition = LocalPenalization(self.model)
+
+        # Everything done in log space so addition here is same as multiplying acquisition with local penalization
+        # function.
+        acquisition = self.acquisition + local_penalization_acquisition
+
+        x_batch = []
+        for i in range(self.batch_size):
+            # Collect point
+            x_next, _ = self.acquisition_optimizer.optimize(acquisition)
+            x_batch.append(x_next)
+
+            # Update local penalization acquisition with x_next
+            f_min = np.min(self.model.Y)
+            lipschitz_constant = _estimate_lipschitz_constant(self.parameter_space, self.model)
+            local_penalization_acquisition.update_batches(np.concatenate(x_batch, axis=0), lipschitz_constant, f_min)
+        return np.concatenate(x_batch, axis=0)
+
+
+def _estimate_lipschitz_constant(space: ParameterSpace, model: IDifferentiable):
+    """
+    Estimate the lipschitz constant of the function by max norm of gradient currently in the model. Find this max
+    gradient norm using an optimizer.
+    """
+    def negative_gradient_norm(x):
+        x = np.atleast_2d(x)
+        d_mean_d_x, _ = model.get_prediction_gradients(x)
+        result = np.sqrt((np.square(d_mean_d_x)).sum(1))  # simply take the norm of the expectation of the gradient
+        return -result
+
+    # Evaluate at some random samples first and start optimizer from point with highest gradient
+    random_design = RandomDesign(space)
+    samples = random_design.get_samples(N_SAMPLES)
+    samples = np.vstack([samples, model.X])
+    gradient_norm_at_samples = negative_gradient_norm(samples)
+    x0 = samples[np.argmin(gradient_norm_at_samples)]
+
+    # Run optimizer to find point of highest gradient
+    res = scipy.optimize.minimize(negative_gradient_norm, x0, bounds=space.convert_to_gpyopt_design_space().get_bounds(),
+                                  options={'maxiter': MAX_ITER})
+    lipschitz_constant = -res.fun[0]
+
+    if lipschitz_constant < 1e-7:
```

Comment:
Both this number and the `10` below should go into variables with clear names, for readability if nothing else

Reply:
Done